File: ruby/easy/house_robber.rb
# @param {Integer[]} nums
# @return {Integer}
def rob(nums)
    memo = [0]
    nums.each do |num|
        memo.push(num)
    end
    for i in 2..memo.length-1
        memo[i] = [memo[i-1], nums[i-1] + memo[i-2]].max
    end
    return memo[memo.length-1]
end

File: javascript/easy/houseRobber.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {
    return maxRob(0, nums, {});
};

var maxRob = function(i, nums, memo) {
    if(i >= nums.length) { 
        return 0; 
    } else if(memo[i] == null) {
        memo[i] = Math.max((nums[i] + maxRob(i+2, nums, memo)), maxRob(i+1, nums, memo));
    }
    return memo[i];
}